internal/os_updater/ubuntu.go

```diff
@@ -19,22 +27,211 @@ type UbuntuDownloader struct {
+	default:
+		return uint64(parsedSize)
+	}
 }
 
-// UbuntuRebooter is the concrete implementation of the IUpdater interface
+const noUpdateAvailable = "0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded."
+
+func getEstimatedSizeInBytesFromAptGetUpgrade(upgradeOutput string) (bool, uint64, error) {
+	log.Printf("Apt-get upgrade output: %s", upgradeOutput)
+	var outputLines []string
+	for _, line := range strings.Split(upgradeOutput, "\n") {
+		if strings.Contains(line, "After this operation,") {
+			outputLines = append(outputLines, line)
+		} else if strings.Contains(line, noUpdateAvailable) {
+			// No update available.  System is up to date
+			return false, 0, nil
+		}
+
+	}
+	output := strings.Join(outputLines, "\n")
+
+	updateRegex := regexp.MustCompile(`(\d+(?:,\d+)*(\.\d+)?)(\s*(kB|B|MB|GB)).*(freed|used)`)
+	matches := updateRegex.FindStringSubmatch(output)
+
+	if matches == nil {
+		return false, 0, fmt.Errorf("failed to get size of the update")
+	}
+
+	freedOrUsed := matches[5]
+
+	if freedOrUsed == "used" {
+		sizeString := strings.Replace(matches[1], ",", "", -1)
+		return true, sizeToBytes(sizeString, matches[4]), nil
+	}
+
+	log.Println("Update will free some size on disk")
+	return true, 0, nil
+}
+
+func noDownload(packages []string) [][]string {
+	log.Println("No download mode")
+	cmds := [][]string{
+		{"dpkg", "--configure", "-a", "--force-confdef", "--force-confold"},
+		{"apt-get", "-o", "Dpkg::Options::='--force-confdef'", "-o",
+			"Dpkg::Options::='--force-confold'", "-yq", "-f", "install"},
+	}
+
+	if len(packages) == 0 {
+		cmds = append(cmds, []string{"apt-get", "-o",
+			"Dpkg::Options::='--force-confdef'", "-o",
+			"Dpkg::Options::='--force-confold'",
+			"--with-new-pkgs", "--no-download",
+			"--fix-missing", "-yq", "upgrade"})
+	} else {
+		cmds = append(cmds, [][]string{append([]string{"apt-get", "-o",
+			"Dpkg::Options::='--force-confdef'", "-o",
+			"Dpkg::Options::='--force-confold'",
+			"--no-download", "--fix-missing", "-yq",
+			"install"}, packages...)}...)
+	}
+
+	return cmds
+}
+
+func downloadOnly(packages []string) [][]string {
+	log.Println("Download only mode")
+
+	cmds := [][]string{
+		{"apt-get", "update"},
+		{"dpkg-query", "-f", "'${binary:Package}\\n'", "-W"},
+	}
+
+	if len(packages) == 0 {
+		cmds = append(cmds, []string{"apt-get", "-o",
+			"Dpkg::Options::='--force-confdef'", "-o",
+			"Dpkg::Options::='--force-confold'",
+			"--with-new-pkgs", "--download-only",
+			"--fix-missing", "-yq", "upgrade"})
+	} else {
+		cmds = append(cmds, [][]string{append([]string{"apt-get", "-o",
+			"Dpkg::Options::='--force-confdef'", "-o",
+			"Dpkg::Options::='--force-confold'", "--download-only",
+			"--fix-missing", "-yq", "install"}, packages...)}...)
+	}
+
+	return cmds
+}
+
+func fullInstall(packages []string) [][]string {
+	log.Println("Download and install mode")
+
+	cmds := [][]string{
+		{"/usr/bin/apt-get", "update"},
+		{"dpkg-query -W -f='${binary:Package}\\n'"},
```

Comment:
does this need to get sliced into individual strings as in `downloadOnly` above? or can it stay with spaces between the parameters? usually if you are not calling the shell, you'd slice into individual strings

Reply:
I tried various things, it was difficult to get to work.  This is the only way I could get it to actually work.